FSW/src/util/memUtil.cpp

```diff
@@ -0,0 +1,250 @@
+ *  Copies buffer to long term memory when signaled by timing module.
+ * 
+ * Inputs:
+ *  none
+ *  
+ * Outputs:
+ *  None
+ */
+void scienceMemoryHandling()
+{    
+    if (dataProcessEvent.checkInvoked()){ // checks event status from timing module
+        float currentPhotodiodeVoltage = dataProcessing();
+
+        Serial.print("Current Photodiode Voltage: ");
+        Serial.print(currentPhotodiodeVoltage);
+        Serial.print("\n");
+
+        updateBuffer(currentPhotodiodeVoltage, bufIdx);
+    }
+
+    if (saveEvent.checkInvoked()){
+        saveBuffer(bufIdx);
+    }
+}
+
+/* - - - - - - Helper Functions - - - - - - */
+
+/* - - - - - - updateBuffer - - - - - - *
+ * Belongs to Science Memory Handling Module
+ *  
+ * Usage:
+ *  adds a new sample to the data buffer at specified index
+ *  an existing element at index will be overwritten
+ * 
+ * Inputs:
+ *  dataBuffer - pointer to first element of data buffer
+ *  sample - data to be stored (type must match dataBuffer type)
+ *  index - index to store new data at (pass-by-reference)
+ *  
+ * Outputs:
+ *  none
+ */
+void updateBuffer(float sample, int &index){
+    
+    // check that index is valid
+    if (0 <= index && index < BUFFERSIZE){
+        dataBuffer[index] = sample;
+        index++;
+    
+    } else {
+        Serial.print("WARNING: invalid dataBuffer index ");
+        Serial.print("(Science Memory Handling Module)\n");
+    }
+
+    if (index == BUFFERSIZE){
+        // reset index if we have reached end of buffer
+        index = 0; 
+    }  
+}
+
+/* - - - - - - saveBuffer - - - - - - *
+ * Belongs to Science Memory Handling Module
+ *  
+ * Usage:
+ *  saves the buffer array passed in to a file on the flash module in csv format
+ *  appends timestamp to the end of the file
+ * 
+ * Inputs:
+ *  dataBuffer - pointer to first element of data buffer
+ *  index - index to store new data at (pass-by-reference)
+ *  
+ * Outputs:
+ *  file creation status
+ */
+bool saveBuffer(int &index){
+    
+    // create array to hold data in time ascending order
+    float timeSortArray[BUFFERSIZE];
+    int j = 0; // iterator for sorted array index
+
+    // reorder array so that it is ascending in time
+    for (int i = index; i < BUFFERSIZE; i++){
+        timeSortArray[j] = dataBuffer[i];
+        j++;
+    } 
+
+    // loop back to top of array and store remaining values
+    for (int i = 0; i < index; i++){
+        timeSortArray[j] = dataBuffer[i];
+        j++;
+    }
+
+    /* send sorted array to file on flash memory along with timestamp 
+     * see SerialFlash docs for info on these functions
+     * https://github.com/PaulStoffregen/SerialFlash/blob/master/README.md 
+     */
+
+    // check if file exists
+    int fileIdx; // iterator for loop checking file existence
+    bool fileFlag = true;
+    char filename[] = "scienceFile0.csv";   // null-terminated char array
+    int fileIdxOffset = 11;                 // index of file number
+
+    while (fileFlag){
```

Comment:
Should there be an else statement here, if the file doesn't exist, warn? Do we just lose data in this case, if the file doesn't exist and the buffer is overwritten?

Reply:
I will add that in, good call